lib/plugins/cpuUsageThrottle.js

```diff
@@ -0,0 +1,149 @@
+
+    if (opts.err !== undefined && opts.err !== null) {
+        assert.ok(opts.err instanceof Error, 'opts.err must be an error');
+        assert.optionalNumber(opts.err.statusCode, 'opts.err.statusCode');
+    }
+
+    var self = {};
+    self._err = opts.err || defaultResponse;
+    self._limit = opts.limit;
+    self._max = opts.max || 1;
+    self._interval = opts.interval || 250;
+    self._halfLife = (typeof opts.halfLife === 'number') ?
+      opts.halfLife : opts.interval;
+    self._timeout = null;
+    assert.ok(self._max > self._limit, 'limit must be less than max');
+
+    // We start at 0 to compensate for the initial spike in CPU usage a server
+    // experiences when starting up.
+    self._ewma = new EWMA(self._halfLife, 0);
+
+    // self._reject represents the % of traffic that we should reject at the
+    // current point in time based on how much over our limit we are. This is
+    // updated on an interval by updateReject().
+    self._reject = 0;
+
+    // updateReject should be called on an interval, it checks the average CPU
+    // usage between two invocations of updateReject.
+    function updateReject() {
+        pidusage.stat(process.pid, function (e, stat) {
+            // If we fail to stat for whatever reason, err on the side of
+            // accepting traffic, while failure probably indicates something
+            // horribly wrong with the server, identifying this error case and
+            // appropriately responding is way outside the scope of this plugin
+            if (e) {
+                // When cpu is NaN, we should never reject
+                stat = { cpu: NaN };
+            }
+
+            // Divide by 100 to match linux's top format
+            self._cpu = stat.cpu / 100;
+            // Update reject with the % of traffic we should be rejecting. This
+            // is safe since max > limit so the denominator can never be 0. If
+            // the current cpu usage is less that the limit, _reject will be
+            // negative and we will never shed load
+            self._reject =
+                (self._cpu - self._limit) / (self._max - self._limit);
+            self._timeout = setTimeout(updateReject, self._interval);
+        });
+    }
+
+    // Kick off updating our _reject value
+    updateReject();
+
+    function onRequest (req, res, next) {
+        // Check to see if this request gets rejected. Since, in updateReject,
+        // we calculate a percentage of traffic we are planning to reject, we
+        // can use Math.random() (which picks from a uniform distribution in
+        // [0,1)) to give us a `self._reject`% chance of dropping any given
+        // request. This is a stateless was to drop approximatly `self._reject`%
+        // of traffic.
+        if (Math.random() >= self._reject) {
+            return next() // Don't reject this request
+        }
+
+        // If we are rejecting the request, log some metadata and return error
+        req.log.trace({
```

Comment:
Instead of logging trace here, you should just put these parameters in a context object on the error, and that way the audit log will automatically emit it. This has the additional benefit of consumers being able to listen to the 'after' event and emit a metric based on the fact that this got throttled.

Reply:
:+1:

Reply:
Actually, this module is miss-using errors. I should be creating a new error on every rejected request instead of returning the same error. Wil.l have to change that if I want to pass a context through.